spec/derived_build_code_formatter_spec.rb

```diff
@@ -96,25 +208,108 @@
-        expect { described_class.new(prefix: '-1') }.to raise_error(/Prefix must be a single digit or empty string/)
+      it 'rejects non-string and non-integer prefix types' do
+        # Arrays, hashes, and other objects
+        expect { described_class.new(prefix: []) }.to raise_error(/Prefix must be a string or integer/)
+        expect { described_class.new(prefix: {}) }.to raise_error(/Prefix must be a string or integer/)
+        expect { described_class.new(prefix: Object.new) }.to raise_error(/Prefix must be a string or integer/)
+        expect { described_class.new(prefix: true) }.to raise_error(/Prefix must be a string or integer/)
+        expect { described_class.new(prefix: false) }.to raise_error(/Prefix must be a string or integer/)
+        expect { described_class.new(prefix: 3.14) }.to raise_error(/Prefix must be a string or integer/)
       end
+    end
+  end
 
-      it 'rejects symbols and special characters' do
-        expect { described_class.new(prefix: '@') }.to raise_error(/Prefix must be an integer digit/)
-        expect { described_class.new(prefix: '#') }.to raise_error(/Prefix must be an integer digit/)
+  describe 'version component validation' do
+    context 'with version components within digit limits' do
+      it 'accepts version components that fit within their digit limits' do
+        formatter = described_class.new(major_digits: 1, minor_digits: 2, patch_digits: 2, build_digits: 3)
+
+        # Valid cases: major=9 (1 digit), minor=99 (2 digits), patch=99 (2 digits), build=999 (3 digits)
+        version = Fastlane::Models::AppVersion.new(9, 99, 99, 999)
+        expect { formatter.build_code(version: version) }.not_to raise_error
+      end
+
+      it 'accepts version components at exactly the digit limit' do
+        formatter = described_class.new(major_digits: 2, minor_digits: 1, patch_digits: 3, build_digits: 2)
+
+        # Edge cases: exactly at limits
+        version = Fastlane::Models::AppVersion.new(99, 9, 999, 99)
+        expect { formatter.build_code(version: version) }.not_to raise_error
+      end
+    end
+
+    context 'with version components exceeding digit limits' do
+      it 'rejects major version exceeding digit limit' do
+        formatter = described_class.new(major_digits: 1, minor_digits: 2, patch_digits: 2, build_digits: 2)
+        version = Fastlane::Models::AppVersion.new(10, 1, 1, 1) # major=10 is longer than 1 digit
+
+        expect { formatter.build_code(version: version) }.to raise_error(
+          /Version component value \(10\) exceeds maximum allowed width of 1 characters/
+        )
+      end
+
+      it 'rejects minor version exceeding digit limit' do
+        formatter = described_class.new(major_digits: 2, minor_digits: 1, patch_digits: 2, build_digits: 2)
+        version = Fastlane::Models::AppVersion.new(1, 23, 4, 5) # minor=23 > max(9) for 1 digit
```

Comment:
I initially suggested to update all those comments but I see you only updated the first one 🙃 

Adding GitHub suggestions for all of them below to make it easier for you to apply them in batch directly from the GitHub UI
```suggestion
        version = Fastlane::Models::AppVersion.new(1, 23, 4, 5) # minor=23 is longer than 1 digit
```